klippa/src/serialize.rs

```diff
@@ -0,0 +1,394 @@
+//! serializer
+use fnv::FnvHashMap;
+use write_fonts::types::Scalar;
+
+#[derive(Clone, Copy, Debug)]
+#[allow(dead_code)]
+pub(crate) struct SerializeErrorFlags(u16);
+
+#[allow(dead_code)]
+impl SerializeErrorFlags {
+    pub const SERIALIZE_ERROR_NONE: Self = Self(0x0000);
+    pub const SERIALIZE_ERROR_OTHER: Self = Self(0x0001);
+    pub const SERIALIZE_ERROR_OFFSET_OVERFLOW: Self = Self(0x0002);
+    pub const SERIALIZE_ERROR_OUT_OF_ROOM: Self = Self(0x0004);
+    pub const SERIALIZE_ERROR_INT_OVERFLOW: Self = Self(0x0008);
+    pub const SERIALIZE_ERROR_ARRAY_OVERFLOW: Self = Self(0x0010);
+}
+
+impl Default for SerializeErrorFlags {
+    fn default() -> Self {
+        Self::SERIALIZE_ERROR_NONE
+    }
+}
+
+impl PartialEq for SerializeErrorFlags {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl std::ops::BitOr for SerializeErrorFlags {
+    type Output = Self;
+
+    /// Returns the union of the two sets of flags.
+    #[inline]
+    fn bitor(self, other: SerializeErrorFlags) -> Self {
+        Self(self.0 | other.0)
+    }
+}
+
+impl std::ops::BitOrAssign for SerializeErrorFlags {
+    /// Adds the set of flags.
+    #[inline]
+    fn bitor_assign(&mut self, other: Self) {
+        self.0 |= other.0;
+    }
+}
+
+impl std::ops::Not for SerializeErrorFlags {
+    type Output = bool;
+    #[inline]
+    fn not(self) -> bool {
+        self == SerializeErrorFlags::SERIALIZE_ERROR_NONE
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Default)]
+enum OffsetWhence {
+    #[default]
+    Head,
+    Tail,
+    Absolute,
+}
+
+// the maximal len of vec in the object_pool is not expected to exceed u32::MAX
+const NULL_POOL_IDX: usize = u32::MAX as usize;
+
+type PoolIdx = usize;
+#[allow(dead_code)]
+struct Object {
+    head: usize,
+    tail: usize,
+    real_links: Vec<Link>,
+    virtual_links: Vec<Link>,
+    next_obj: PoolIdx,
+}
+
+impl Default for Object {
+    fn default() -> Object {
+        Object {
+            head: 0,
+            tail: 0,
+            real_links: vec![],
+            virtual_links: vec![],
+            next_obj: NULL_POOL_IDX,
+        }
+    }
+}
+
+#[allow(dead_code)]
+#[derive(Default)]
+enum LinkWidth {
+    #[default]
+    Two,
+    Three,
+    Four,
+}
+
+type ObjIdx = u32;
+#[allow(dead_code)]
+const INVALID_OBJ_IDX: ObjIdx = ObjIdx::MAX;
+
+#[allow(dead_code)]
+#[derive(Default)]
+struct Link {
+    width: LinkWidth,
+    is_signed: bool,
+    whence: OffsetWhence,
+    bias: u32,
+    position: u32,
+    objidx: ObjIdx,
+}
+
+#[derive(Default)]
+#[allow(dead_code)]
+pub(crate) struct Serializer {
+    pub(crate) start: usize,
+    pub(crate) end: usize,
+    pub(crate) head: usize,
+    pub(crate) tail: usize,
```

Comment:
I'd document what these describe. I'm also slightly cautious that these are `pub(crate)`, since I assume we would be violating some invariants if somebody changed them out from under the serializer?

Reply:
Okay so after reading the rest of this:

I wonder if it makes sense to have some kind of buffer type that lives outside of the serializer, and which manages just the buffer? this could then be managed by the serializer?

It feels this is a reasonably isolated task, and would reduce overall complexity somewhat?